### DISCUSSION TITLE
Feature Request - new operator keyword: is not

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#27] [General] [answered]
Consider adding a new operator keyword "isnt" or "isnot" which acts as the converse of "is".

"is" is used like this:

    if (obj is MyObject)  
       {  
       /* do stuff */
       } 

coding the converse must be done currently like this:

    if (! obj is MyObject)  
       {  
       /* do stuff */
       } 

whereas this is much easier on the eye:

    if (obj isnt MyObject)  
       {  
       /* do stuff */
       }

**Comment by The-Futurist** [comment]
I see this isnt the place to submit feature requests, clsoing this.

**Comment by gafter** [comment]
@Korporal This is now the place to open discussion on feature requests, so you can reopen this if you want.

**Comment by The-Futurist** [comment]
As suggested, I'm reopening this.

**Comment by jnm2** [comment]
I like `isnot` or `isnt` and I dislike `!is` which has been suggested elsewhere. So many times I have to backtrack and prepend a `!` and insert parentheses. This would be one of those day-to-day coding improvements for me. :)

**Comment by The-Futurist** [comment]
Exactly, if the designers felt is necessary to add "is" supporting its converse seems like a natural step.

**Comment by mattwar** [comment]
aint?

**Comment by Thaina** [comment]
I would like `is!` more than `isnot`

And I would against `isnt`

Please don't add more keyword

**Comment by MgSam** [comment]
It's really cringe inducing watching Mads' video on C# 7.0 and seeing `if (!(o is int i)) return;` as an example of a *good* coding pattern. `isnot` (or whatever you want to call it- I don't really care about the syntax) is painfully necessary.

Unrelated-
@Korporal Can you rename this to include the words "is not" in the title? It's hard to search for this issue as its currently written.

**Comment by gulshan** [comment]
Related- [Negetive pattern](https://github.com/dotnet/roslyn/issues/16766) by @alrz

**Comment by phi1010** [comment]
@Thaina: `variable is ! Type ` might be confusing, people might easily mistake it for `variable is Type ! ` (Non-Nullable-Reference) or might think it is used as an operator returning a boolean (or whatever an overload returns).

**Comment by Thaina** [comment]
@phi1010 In my case I was differentiate between `is !` and `is!`
What I mean is let `is!` **without space** as an operator by itself

`is!` **without space** OK?
Why you add space to make confusion anyway?
You surely be confused if you add space between it

What I propose is not to separate `is` and `not` as multiple operator
I propose that we should have operator `is!`
like `==` also have operator `!=`
You cannot have operator `! =` you must have operator `!=` **without space**

In many language they add operator by combine multiple symbol **without space**. Such as javascript `===` and `!==` symbol in addition to `==` and `!=`
They never make a bad choice such as `== not` or `not ==` or `! ==` They must add `!==` **without space** so it never be a confusion

Unlike what you try to do `is not` and `is !`

**Comment by HaloFour** [comment]
@Thaina 

Whitespace is an awful delineation of syntax.  Not everyone programs using monospace fonts or with whitespace glyphs enabled.

**Comment by Thaina** [comment]
@HaloFour I agree with you but the root cause itself came from Hejlsberg himself to select `is` as an operator. If we have symbol operator for `is` then put `!` to it would be easy

I always feel wrong that we have word as operator instead of symbol. But now it's too late to argue about that

But in the same sense. While we cannot write `varAisType`. We must write `varA is Type` We must have whitespace there for ` is `. Then it should be the same as `is!` and must write `varA is! Type`
In any font it will be the same whitespace so if anyone could confuse by the space of font he use for ` is ` then something must be wrong there

**Comment by gafter** [comment]
What is the relationship to #246 ?

**Comment by phi1010** [comment]
@Thaina: The way the language is currently structured, I expect to be able to insert additional spaces at any position between two lexical tokens (literals, operators made of special characters, and words such as keywords, variables, method names, etc.).

It is somewhat obvious that it is not a good idea to separate `a++;` into `a+   +;`, but I (and maybe also silly text editors) expect to be able to insert spaces between letters and special characters. If there is currently a situation in C# where inserting spaces between letters and special chars is already invalid, please correct me, I can't think of any example.

**Comment by Thaina** [comment]
@phi1010 That's why I propose that we should have operator `is!` not to allowing  !  operator after `is`

**Comment by Thaina** [comment]
@phi1010 You cannot write `a = = b` it is syntax error

**Comment by jnm2** [comment]
@Thaina He's saying, show any example where it changes the meaning to insert whitespace between a letter character and a punctuation character. And I agree with his point that it's a desirable aesthetic to preserve.

**Comment by Thaina** [comment]
@phi1010 @jnm2 I think both of you are wrong to consider that `is` is a letter

`is` is the operator keyword and whenever it combine with any other operator then it must be considered as an operator so you can't put any operator separated

In javascript you cannot wrote `a ! == b ` or  `a != = b ` it is syntax error and will not make your code run

**Comment by phi1010** [comment]
@Thaina 

"i" is a letter, "s" is a letter, "!" is a special character.

**Comment by MgSam** [comment]
It's a shame this issue has devolved into a syntax war rather than discussing the underlying feature, which IMO is the important part.

**Comment by jnm2** [comment]
Keywords should stay letters-only. A bunch of us think that anything else is ugly and ambiguous to read. `is` is a keyword which comprises only letters, and its mirror complement should be the same.

@MgSam This proposal calls for both discussions. So... what would you like to discuss? :smile:

**Comment by CyrusNajmabadi** [comment]
> It's a shame this issue has devolved into a syntax war rather than discussing the underlying feature, which IMO is the important part.

I don't see how that's hte case.  If this is about the 'underlying feature', then there is no need for a change here as C# already supports the scenario being discussed.  The primary reason this discussion exists is *precisely* because the *syntax* that C# has for supporting this is somewhat ungainly and people would like a more ergonomic syntax to accomplish the same thing.  As such, this discussion really does boil down to "is there a syntax that feels like it would substantially improve things enough to merit inclusion".

**Comment by alrz** [comment]
With https://github.com/dotnet/csharplang/issues/35 we are practically starting to see operator-keyword combinations as language constructs. So I think `!is` or `is!` are viable options here.

>What is the relationship to #246 ?

#246 suggests a recursive pattern to negate `is`, but it would require complex/inconsistent definitive assignment rules. So I think this proposal is a good alternative to `if (!(e is P))` if we're going to introduce a sugar just for that.

**Comment by MgSam** [comment]
@CyrusNajmabadi My point was that 99% of the value here is in avoiding the parenthesis mess and inversion of readability that comes with `if(!(x is condition))`. 

Whether the syntax is `is!`, `!is`, `isnot`, `is not`, `isnt`, `aint`, or `hejlsberg` seems trivial. It smells very strongly like the ridiculous `private protected` arguments of yore. 

Yet people are downvoting what seems like, in principle, a no-brainer addition because of these silly minor syntax differences.

**Comment by CyrusNajmabadi** [comment]
> silly minor syntax differences.

'Minor' is in the eye of the beholder.  Syntax matters a lot to many developers.  It's what you're working with day in and day out.  Because this feature is all about providing a better syntax to *avoid the current syntax problem*, people are of course going to be very interested, passionate, and opinionated on what syntax we would use.

**Comment by MgSam** [comment]
I agree, many people are passionate about it, but IMHO their priorities are misplaced. You'll get used to whatever keyword is chosen in a day or two, whereas if the feature is passed over altogether because of these disagreements everyone loses. 

`private protected` being the example- everyone agreed the feature is useful, and it was punted two whole language versions back (at least 4 years) just because of petty disagreements on the keywords. And the end result? The LDM is just going to use the same keywords they wanted to to start with.

So, IMO, the more important discussion is "Is some kind of `isnot` operator (regardless of exact syntax) worthwhile?" If the answer to that is yes, then let the LDM vote internally and chose whatever word(s) they like best. You're never going to get the larger community to agree on keyword choices.

**Comment by The-Futurist** [comment]
Well I agree with that too, almost anything is better than the clumsy !(X is string) BS. Once again we see how clinging to the feeble, dated C syntax continues to hamper the elegance of modern languages. Frankly the designers of C# should have ditched C, Java and C++ syntax and actually designed a new language with a better syntax.

**Comment by phi1010** [comment]
@MgSam: I don't think one can get easily used to an keyword that is not consistent with the language itself. If "is!" is the only keyword that is uses this combination, it just feels weird. If this is consistently used in the future by "await?" or something else, it won't be anymore.

Also, there still is the variant "is not": That's two keywords -- trying to be similar to natural language. In my opinion, resembling natural speech is significantly harder to write, since the right word order can't be easily determined. It might be easier to read in statements that are already easier to read, but when used in complex statements, it can quickly become very hard to read. I'd like to keep using simple standalone keywords, such as `isnot` instead of `is not`, even if this causes more keywords to be necessary. I fear we will get a unreadable syntax with an unpredictable keyword order otherwise, if the code was written badly.

Python provides examples of how unreadable it might become:
`(1 if y is not x else 2  for y in x  for z in y  where z > 1)`
Linq does not try to mimic human speech, thus -- in my opinion -- it is readable a lot better, even when formatted badly:
`from y in x from z in y where z > 1 select y != x ? 1 : 2`

**Comment by MgSam** [comment]
@phi1010 
I happen to prefer `isnot` myself, but for the most part I believe this debate is [bike shedding](https://en.wikipedia.org/wiki/Law_of_triviality).

**Comment by aluanhaddad** [comment]
Yay, 🚲:house: time!
I would be in favor of `isnot` but against `is not`, because two token keywords are very rare in C#, against `is!` because it looks like two tokens but is one, and against `isnt` because it feels unprofessional (I keep wanting to add the missing apostrophe 😝). @MgSam's `hejlsberg` keyword is definitely worth considering as well.

**Comment by BreyerW** [comment]
There is already precedence with `else if` though, so, personally, i think for consistency it should be `is not` even if it isnt perfect. (`nameof`, `typeof` and the like are more like methods to me especially since they require `()`).

In short i vote for `is not`

**Comment by aluanhaddad** [comment]
`else if` is just a line formatting convention the `else` token binds to the preceding `if` and the `if` token introduces an independent conditional in the scope of the `else`.

**Comment by weitzhandler** [comment]
I also find the need for parenthesis when negating pretty annoying. Would love to see a syntactic sugar addressing this.

**Comment by tauheedul** [comment]
I'm happy with isnot. It reads well and vb.net has had it for a while

**Comment by renatoeufe** [comment]
+1 for isnot

**Comment by mshwf** [comment]
Yes, let's Englishify C#

**Comment by binoypatel** [comment]
+1 for isnot

**Comment by jvlppm** [comment]
I would prefer support for:

if !(foo is Bar) { }

But one thing does not exclude the other.

**Comment by CabuxaMapache** [comment]
+1 for isnot

**Comment by jjvanzon** [comment]
I am OK with isnot, not ecstatic.

@mshwf

> Yes, let's Englishify C#

At the risk of starting a completely different discussion: I wrote about a million lines of VB code and loved it, before I switched to C# and was sold on how it was more schematic, because it actually had less words. So I cannot agree with 'Yes, let's Englishify C#'. The way C# looks more like a diagram compared to VB: that's what I had been missing in VB all those years.

**Comment by mshwf** [comment]
@janjoostvanzon first off excuse my English (it is not my first language :D)
I didn't use VB, but I saw some code, what a mess!I don't know is that because I'm not familiar with it like C# or because it really is?!
but the beauty of C# is that it's syntax is put in the right use: English when needed and when express in words what symbols can't, and symbols when it can say a lot.
but combining the two approaches in one single statement is kind of weird especially when we use it in wordy taste like `is` , `(foo isnt Bar)` is clearer than `!(foo is Bar)` 
my comment may not be precise, otherwise we will fall into another VB syntax, the precise comment must be: **let's Englishify what must be Englishified.**

**Comment by lachbaer** [comment]
FYI, I'm currently [prototyping](https://github.com/lachbaer/roslyn/tree/isnotkeyword) this feature and I am using `isnot` as the keyword.

~~The only problem that can arise with `isnot` is that somebody has a variable defined with that name.~~
```cs
var isnot = CheckOperationForNegation(operation);
```
But a first quick search on code with 'isnot' hasn't found anything.
**Update:** implemented as contextual keyword.

**Comment by yaakov-h** [comment]
@lachbaer can't you prototype it as a contextual keyword, e.g. how `async async async(async async)` is still valid C#?

**Comment by lachbaer** [comment]
@yaakov-h ~~You, mean to be contextual only in expressions? Maybe in a second step. To be honest, I currently don't know how to do that and must have a thorough look at that first. Update: trying...~~  Update: done 😃

**Comment by weitzhandler** [comment]
@janjoostvanzon VB used to be my native language before C# took over, and indeed the `IsNot` is a very convenient keyword. The idea of lots of parentheses in C# (especially in conversion), can get really ugly.  `isnot` has a potential of reducing a bit the `if(!(value is IItem item))` double-parenthesis overhead.

**Comment by lachbaer** [comment]
I think it is safe to disallow a variable declaration, discard pattern and var pattern in `isnot` patterns and allow all other constructs.
```cs
if ( obj isnot Polar coords ) ...    // error, no declaration patterns allowed in isnot operator
if ( obj isnot _ ) ...    // error, no discard pattern allowed in isnot operator
if ( obj isnot var coords ) ...    // error, no var pattern allowed in isnot operator

if ( obj isnot Polar(double X, double Y) ) ...    // allowed
```
The latter should be allowed because it allows for correct pattern matching and the variable *names* can make clear to the reader what concrete pattern we are checking. Of course no additional variable shall be created by the compiler.

Alternatively the errors can be warnings, because they make no sense together with `isnot`.

**Comment by jnm2** [comment]
`if (obj isnot Polar coords)` should be legal because `if (!(obj is Polar coords))` is legal.
This is becoming one of the primary reasons I want `isnot`. :-)

**Comment by jnm2** [comment]
> This is becoming one of the primary reasons I want isnot. :-)

@DavidArno is not going to want to hear this 😁

**Comment by lachbaer** [comment]
@jnm2 But that doesn't make sense? 'coords' is scoped to the if-block that is never executed. Also a not user-defined conversion from obj to Polar must exist and 'operator is' is only questioned on parametered patterns.

**Comment by DavidArno** [comment]
@jnm2,

Why are you wanting to do `if (!(obj is Polar coords))`, much less `if (obj isnot Polar coords)`? Does that mean you are using `coords` in the `else` clause? 😱

**Comment by jnm2** [comment]
@lachbaer Actually it is not scoped to an if block. It's scoped as though it was declared the statement before the if. See below. There was a huge [discussion](https://github.com/dotnet/roslyn/issues/14697#issuecomment-262404116) on this last year. I was skeptical, but I've ended up appreciating the wider scope.

@DavidArno In almost every case there is no else clause but the code looks like this:

```c#
if (!(obj is Foo foo)) return; // Or continue or break or throw
//  Use foo
if (!(obj2 is Polar coords)) return;
// Use coords
// ...
```

**Comment by lachbaer** [comment]
According to the current version of the specs in this repo the scope is the block only. The current master branch scope it to the if enclosing block however. In case of 'isnot' the variable then shalll be null, correct?

**Comment by jnm2** [comment]
> In case of 'isnot' the variable then shalll be null, correct?

Right, `default(T)`, same as `!(obj is Foo foo)`.

**Comment by DavidArno** [comment]
@jnm2,

Oh, I see why you say I wouldn't like it. I still think it sad the team went for scope leakage to make those out vars work, rather than allowing `return` etc to be expressions like `throw`:
```cs
var validatedFoo = obj is Foo foo : foo ? return;
```
The above is far more expressive than weird scoping rules in my view. But we lost that argument, so I guess you might as well use what we did get.

**Comment by jnm2** [comment]
@DavidArno I strongly dislike having to say come up with two names (`foo` and `validatedFoo`) for two locals whose values are completely identical. I would end up writing `var foo = obj is Foo _ ? _ : return;` and then getting mad that I can't use `_` there and writing `var foo = obj is Foo f ? f : return;`.

`var foo = obj as Foo ?? return;` now that I'd use. But that doesn't work with `dict.TryGetValue`, so I'd still want the wider scope for that out var.

**Comment by aluanhaddad** [comment]
I don't want to use it in that way at all but, as you say, now that it's here you might as well take advantage of it.

**Comment by lachbaer** [comment]
I was lately thinking about whether `is not` isn't actually a better choice. The reason being `switch` (and *match*) pattern matching.
```cs
switch (obj) {
    case Euro(var amount): [...]
    case USDollar(var amount): [...]
    case not LocalCurrency: [...]    // it's not Euro, nor USDollar, but is a currency: skip this block
    case decimal: [...]
    default: [...]
}
```
The `not` would now belong to the pattern itself. This example could be written differently by reordering the `case` blocks, but like with all discussions about negation (`isnot`, `ifnot`, `until`) it might ease expressing the wishes, especially when the `case` is appended by a `when` and if it gets a bit more complex.
(/cc @gafter)

**Comment by tuespetre** [comment]
@lachbaer I was reading this thread all the way down surprised that nobody had yet suggested a simple `not`. You are the winner 🏅 🏆

**Comment by aluanhaddad** [comment]
@lachbaer although that suggestion implies that there would then be two tokens that mean "not", we already now have two ways of saying "is null", `n == null` and `n is null`, so I find your argument compelling.

**Comment by BreyerW** [comment]
I like standalone `not` keyword. Internally it could be represented as negated `is` (where applicable) i.e.:

`if(object not null)` transformed to `if(!(object is null))` after compilation. I guess similar transformation could be applied to `loop`s and `if`s.

This way you would have to define `is` operator only once and it will work immediately for both `is` and negated version.

**Comment by aluanhaddad** [comment]
@BreyerW If `not` negates, it would be
```C#
if (object is not null) { ... }
```

**Comment by lachbaer** [comment]
It is important to point out that `object is not null` is still a classical operation on operator `is`, however with a *negated* pattern. The advantage of this is that the negation can be used everywhere where a pattern is allowed (also in far future additions).

**Comment by BreyerW** [comment]
@aluanhaddad  What you present would probably require new operator. Even if not (I guess `is not` could be transformed same way) `not` alone is still shorter and a bit easier to parse (no extra space which can be irritating obstacle).

@lachbaer well same apply for just `not` since it could be equivalent to `!(is)` after compilation step.

**Comment by aluanhaddad** [comment]
This proposal is for a negated version of `is` via some keyword: `is not`, `isnot`, `isnt`, `aint`. @lachbaer showed an example of using `not` as a separate `keyword` to negate `case` as well. That seems interesting.

> not alone is still shorter and a bit easier to parse

Use `if (x != null)`.

Being able to write
```C#
if (x isnot null) { ... }
```
would be a _consequence_ of adopting this proposal in any form, but I do not think it is the `goal`.

**Comment by BreyerW** [comment]
Well `null` was just example. I argued that `not` could be another alternative for all those already mentioned. Thats all.

At the end, it is up to designers to pick one, if ever (although i would like to see this since I had few times were i would like to make negation in less verbose manner).

**Comment by CyrusNajmabadi** [comment]
> It is important to point out that object is not null is still a classical operation on operator is, however with a negated pattern. The advantage of this is that the negation can be used everywhere where a pattern is allowed (also in far future additions).

I like this.

**Comment by mattwar** [comment]
Negating patterns in general is kind of awkward and probably only makes sense for simple value patterns.

Given, t = (object, object)

```cs
is (int x, string y) 
```
When the expression is true, both x and y are definitely assigned.
When the expression is false, neither x nor y are definitely assigned.

```cs
is not (int x, string y)
```

When the expression is true, neither x nor y is definitely assigned.
When the expression is false, neither x nor y is definitely assigned.

This happens, because there are many ways to fail matching the pattern but only one way to succeed. You can fail if the type doesn't match, in this case not a two element tuple, of if the first element is not an integer or if the second element is not a string.

**Comment by lachbaer** [comment]
@mattwar 
The naming of the parameter can even help in an expression where they aren't assigned any value, just for the sake of verbosity on what 'int' and 'string' shall represent. Also there might be further null checks on those parameters later in the code or the variables default value is returned by the function. So I even see potential use for them in negated expressions.

**Comment by mattwar** [comment]
@lachbaer you cannot use the variables in the code if they are never definitely assigned.

**Comment by lachbaer** [comment]
Well, there are two possibilities. 1st, ignore the identifiers completely ; 2nd assign the default value in case of a (negative) match. Btw, what happens with the variables of a standard pattern when there is no match and the scope is the enclosing block?

**Comment by gafter** [comment]
@mattwar I think the intent is that the negated pattern is precisely the negation of the normal pattern, so that for the expression `e is not (int x, string y)` both x and y are definitely assigned when false.

**Comment by mattwar** [comment]
@gafter you are right.  My mind was boggled by the negation and I couldn't think clearly.

**Comment by lachbaer** [comment]
@gafter 
> ` e is not (int x, string y) both x and y are definitely assigned when false.`

Assigned to default values or assigned to the same values as if `e is (int x, string y)`? Because it can be seen both ways, either `is not` is the *exact* equivalent of `!(. is .)` - i.e. the variables are assigned as if the pattern was matching - or it is an expression of its own with a different outcome than the simple negation of the positive pattern, i.e. the variables are assigned with their type default values when the `is not ...` expression is true, and assigned with their deconstructed values when `is not ...` is false.

The latter makes a bit more sense to me, because that way `is not` is *not* ( 😀 ) just another way to write `!(. is .)` but has its own meaning. But you're deeper into that topic and may have different arguments?

**Comment by qaqz111** [comment]
Here's a list of possible usages of `not`:

# 1. if, while:
```C#
if not(xxx){}

while not(xxx){}

do {} while not(xxx);
```
There are many discussions of this feature:

- [#157: Proposal: `not` keyword for `if` and `while`](https://github.com/dotnet/csharplang/issues/157)
- [#560: Proposal: Add 'Unless' conditional statement, as seen in Ruby](https://github.com/dotnet/csharplang/issues/560)
- [#562: Poll: `until` in addition to `while`](https://github.com/dotnet/csharplang/issues/562)
- [#568: Proposal/Discussion: Negation of if/while Condition](https://github.com/dotnet/csharplang/issues/568)

Basically I vote for [HaloFour's opinion in #568](https://github.com/dotnet/csharplang/issues/568), allowing negation prefix operator `!` to appear outside of the required parenthesis, this looks clean and simple. But if other usages make `not` a necessary keyword, I think it is not a big problem to let `not` work with `if` and `while` too.

-----
# 2. is:
```C#
if (a is not xxx){} //-> if not(a is xxx){}, same to `while`

var notok = a is not xxx;
```
Actually as you can see, `is not` can be converted to `if not` or `while not` when used with `if` or `while`, but the `not` is still necessary when you just want to store the result to a variable.

-----
# 3. switch:
```C#
switch (obj) {
    case Euro(var amount): [...]
    case USDollar(var amount): [...]
    case not LocalCurrency: [...]    // it's not Euro, nor USDollar, but is a currency: skip this block
    case decimal: [...]
    default: [...]
}
```
- [earlier comment of @lachbaer](https://github.com/dotnet/csharplang/issues/27#issuecomment-302863547)

This is useful when we want to filter a specific derived type among multiple ones.
The equivalent for now is:
```C#
switch(creature) {
    case Animal a when !(a is Cat): [...]

//with `not` we can write this:

    case not Cat: [...]
}
```
The use of `not` effectively reduced keystrokes and the code became more clear.
But weak point still exists: we can filter **only ONE** derived type.
```C#
switch(creature) {
    case Animal a when !(a is Cat || a is Dog || a is Fish): [...]

//with `not`, some other enhancement must be done to filter multiple derived types:

    case not Cat Dog Fish: [...]
//or
    case not Cat, Dog, Fish: [...]
//or
    case not Cat not Dog not Fish: [...]
//or
    case not Cat nor Dog nor Fish: [...]
//or
    case not Cat || Dog || Fish: [...]
//or
    case not Cat | Dog | Fish: [...]
}
```
Looks like the 1st or 2nd solution is most acceptable:
- no repeated same keywords(3rd): a bunch of same `not` appearing in one single line will drive obsessives mad lol :D
- no new keyword(4th): although 'nor' makes statement closer to natural language, but introducing one more new keyword is an expensive cost, not as good as the 3rd solution.
- no new usage of current operator(5th, 6th): someone reads `||` and `|` as 'or', similar to the 4th solution, but the cost is expensive too.

The difference between 1st and 2nd is, the using of comma makes code more intuitive to read.

So, giving up support filter multiple derived types or make another enhancement, we have to face one.

One more question of using `not` with `switch` is, should `not` work with Enums and value types?
```C#
switch(weekday) {
    case DayOfWeek.Monday: [...]
    case not DayOfWeek.Sunday, DayOfWeek.Saturday: [...] //workday besides monday goes here
}
```

```C#
void foo(int number) {
    switch(number) {
        case 10: [...]
        case int k when k > 20: [...] //I have words to say about this `int`... XD
        case not 1, 3, 4: [...] // 2, 5-9, 11-20 goes here
        default: [...] // 1, 3, 4 goes here
    }
}
```

-----
# 4. catch:
```C#
try
{}
catch (AnimalException ex) not CatException
{}

//or, for filtering multiple derived Exception types:

catch (AnimalException ex) not CatException, DogException
{}
```
In my opinion, `try`/`catch` is another specialized switch statement of Exception handling, so if `not` is necessary for `switch`, it should be necessary for `try`/`catch`.

There is a notable question is: the filtered Excetption types **MUST** be derived from which was caught. The compiler can make a check of this.

Think about this:
```C#
//Exception -> AnimalException -> CatException
//Exception -> ArgumentException -> ArgumentNullException

try
{}
catch (ArgumentException ex) not AnimalException
{}
catch (ArgumentException ex) not CatException
{}
```
Since an `ArgumentException` is definitely not an `AnimalException` nor a `CatException`, the `not` clause become meaningless here.

Of course there is equivalent for now, using `when` clause:
```C#
try
{}
catch (AnimalException ex) when (!(ex is CatException))
{}
catch (AnimalException ex) when (!(ex is CatException || ex is DogException))
{}
```

-----
# 5. where (Generic type constraints):
A new idea of `not` or a new usage of `!`, and it seems there is not an equivalent present.
```C#
static class MyExtensions
{
  public static void AsSomething<T>(this IEnumerable<IEnumerable<T>> arg) { }
  public static void AsSomething<T>(this IEnumerable<T> arg) { } where T : not IEnumerable<> // <-- use of not constraint
//or
  public static void AsSomething<T>(this IEnumerable<T> arg) { } where T : !IEnumerable<> // <-- use of ! constraint
}


// Resolves to AsSomething<T>(this IEnumerable<IEnumerable<T>> arg)
new[] { new[] { 1 } }.AsSomething();

// Resolves to AsSomething<T>(this IEnumerable<T> arg) { } where T : IEnumerable<>
new[] { new { FirstName = "Joe", LastName = "Smith" } }.AsSomething(); 
```
- [#707: Allow 'not' generic type constraints](https://github.com/dotnet/csharplang/issues/707)

**Comment by Maxim-Kornilov** [comment]
Hi @qaqz111
Your proposal is more complicated for the compiler and sometimes might be confusing. The simple `not` operator, which would be an opposite of `is` operator, would be enough.

### 1, 2: Conditional operators and loops
`if (x not string)` sounds more natural than `if not(x is string)`. And with `if not` operator you can do nonsense like this: `if not (x not string)`.

### 3. Pattern matching
For pattern matching it is confusing and not helping at all.
``` csharp
object x = 1.0;
switch (x)
{
    case int i: break;
    case not double d: break;
    case not string s: break;
    default: break;
}
```
For the `double` value the execution will continue in the `case not string s` branch and the type of the variable `s` would be `Object`.

### 4. Exception filters
There is no need to invent another syntax for exception filters. Existing syntax will work and look completely fine:
``` csharp
catch (Exception ex) when (ex not OutOfMemoryException)
```

### 5. Generic type constraints
Generic type constraints are helpful when you need to know how you can use a constrained object. How can this constraint help you use variable `obj`?
``` csharp
void Foo<T>(T obj) where T : not String
```

**Comment by JustNrik** [comment]
> 
> 
> Here's a list of possible usages of `not`:
> # 1. if, while:
> 
> ```cs
> if not(xxx){}
> 
> while not(xxx){}
> 
> do {} while not(xxx);
> ```

Why don't we just add do {} until like vb? it's much more understandable IMHO.

```cs
do
{
    // stuff
}while(!(someBoolExpr))

do
{
    // stuff
}until(someBoolExpr)

```

**Comment by CyrusNajmabadi** [comment]
There's already a propsal exploring: `if !(xxx) { }` and `while !(xxx) { }` and `do { } while !(xxx);`

**Comment by The-Futurist** [comment]
> > Here's a list of possible usages of `not`:
> > # 1. if, while:
> > ```cs
> > if not(xxx){}
> > 
> > while not(xxx){}
> > 
> > do {} while not(xxx);
> > ```
> 
> Why don't we just add do {} until like vb? it's much more understandable IMHO.
> 
> ```cs
> do
> {
>     // stuff
> }while(!(someBoolExpr))
> 
> do
> {
>     // stuff
> }until(someBoolExpr)
> ```

Indeed, once again I refer to PL/I

```
do while (condition);
     /// statements
end;

do until (condition);
     /// statements
end;
```

But C# cannot undergo huge grammar changes at this point (no language could).

**Comment by CyrusNajmabadi** [comment]
As mentioned above: 
> There's already a propsal exploring: `if !(xxx) { }` and `while !(xxx) { }` and `do { } while !(xxx);`

I've got a PR on an low-impact impl-strategy, and Aleksey is championing the feature.

**Comment by jongalloway** [comment]
@SleepyBoyy your comment was in violation of our CoC and was deleted. Any such future violations will result in a ban. We value your interaction, as long as it is courteous and productive. If you'd like to discuss, you can contact us at conduct@dotnetfoundation.org

**Comment by sgf** [comment]
This is a bit useful, but not very useful

**Comment by gafter** [comment, marked as answer]
As an extension to the pattern-matching constructs in C# 9.0, we are adding a `not` pattern and a type pattern.  As a result, `is not` will act like a new operator as you request.

See https://github.com/dotnet/csharplang/blob/master/proposals/patterns3.md
